docs/designs/scrapping-workflow-phase-1-design.md

```diff
@@ -0,0 +1,67 @@
+* Workflow can be scaled easily
+* Workflow is observable
+
+**What currently exists**
+* From [this Design Doc](https://www.notion.so/de8ece87473b4f37a0ea037d607d910a?v=1bec41f87a6b42acb639d348d2282195&p=6d6b1797931e465d9cf6f5d4641c911e), Angostura can be used to publish and persist data to google BigQuery.
+* Sambil has been packaging their code as a library: c4v-py (this repo), which can be imported and used elsewhere.
+
+**Proposed design** 
+
+_Goal_: Leverage existing angostura infrastructure as much as possible
+![image](../images/scrapping-workflow-phase-1-design.png)
+
+1. A custom written data uploader processes NGO submitted data and emits the events to the Angostura proxy gate.
+At first, if the input is CSV, this can be a simple csv uploader script.
+
+2. Angostura does it’s processing and persists the data to a big query table
+
+3. A task defined in airflow queries for yet-to-be-enriched data and proceeds to scrape.
+It can then persists the output to the table that classification workflow(s) will start from. 
+   
+    More on this job below. 
+
+* Any further tasks such as cleaning and classification can be done further along the pipeline.
+
+**Scraping task** 
+
+* Here, we can leverage the python_transform job type Angostura offers. (which is backed by the BigQueryPythonOperator). From the Angostura documentation:
+> The python_transform job takes a list of SQL files as arguments, a python function as a string and writes the result of calling that function with the arguments into a target BIgQuery table [...] Each transformation function is a Python function that receives pandas data frames as arguments.
+
+* This means we could write a python module which     
+    * Exposes one function, which takes a data frame as input and produces a data frame as output.
+    * The input consists of the rows of data that need scraping. 
+    * The output can be the same rows, with the scraped data now present in the columns where it should be so that they can be persisted.
+    
+_Then the job becomes_
+1. Query for all the rows of data that haven’t been scrapped yet. 
+   * As a simple example, if there is a nullable field called scraped_data in the persisted data, then the query can simply be select * where scraped_data is null
+   * The query can also be modified to select only rows that contain a specific source, and pass those to a specific scraping task etc. That should make it flexible enough for anything we want to do in phase 2.
+
+2. The python function exposed by the module is called with these rows passed in.
+3. Scraping happens.
+4. Scraping output is persisted.
+ 
+**The python module** 
+
+This can be very simple for the MVP. 
+A sample design can simply be to:
+* Define a scraper interface, which takes in a URL (+ anything else needed) and outputs scraped data.
+* Implement a top level method, which contains a map of URL domains to scraper implementations
+* Output is written to data frame
```

Comment:
```suggestion
* Output is a string
```

To make things even simpler, since the output is just text content, I propose we write a `Str -> Str` interface (URL -> Text Content). That way, if we want a data_frame (or any other kind of Python Object), we can write simple adapters. These are easier to test and decoupled from libraries (and extendable).

Reply:
For example, to tackle the point in [line 35](https://github.com/code-for-venezuela/c4v-py/pull/54/files#diff-9f63de37b91ec12f121941309bec3908e333d6b1288ba31d9fd6b041a07ceaa0R35), if Angostura requires data_frames, we (or Angostura team) can write those adapters without much hassle, test them, and they can become plugins of the `c4v-py` library. Something along the lines of `c4v-py.community.angostura` or `c4v-py.contrib.angostura`.

I've seen this approach being used many times by projects. If a certain package inside the community packages proves itself to be robust or _coreish_ enough, it becomes part of the `core` modules.

Reply:
Yeah I'm fine with that approach 👍🏽 
The one thing I'm thinking about though, is if that top level method takes all the urls, then it would be better if the resulting data structure keeps the relationship of URLs -> scraped_data. 

That way we know what to persist for each URL 

I was thinking we could maybe have an internal object in our lib that is used in this top level loop. 
Then the adapter is just a wrapper around that data frame that implements this interface

wdyt?

Reply:
I would make it the other way, but I like your point on persistence. It's important to discuss that.

Ideally, our "object" should be agnostic of the persistence mechanisms. In fact, it should not be an object at all. After all, it is just data. Objects imply behavior.

As a core library, we have to think about our users. They might (and probably will!) be interested in saving the data as CSV, data frames, XML, JSON, or Excel files. We need flexibility. This is why our data should be a super simple data structure (in the strict _computer science_ of the word).

For example, a (Hash)Map—a dictionary in Python. The keys could be the URLs themselves—or a combination of parameters, such as the and timestamp—or a custom hashed computation. In the case of the values, we would have the scraped data itself. To give an example, another map.

This approach is already similar to how one would interact with a NoSQL database, but we are **only** using _Plain Old Python Data Structures_.

The problem when one creates their own objects, whether or not they use a 3rd-party library inside, is that many libraries know how to deal with Python data structures (e.g. `Dict` or `OrderedDict`), but have no idea about custom objects. For instance, `CodeForVenezuelaData` or `CodeForVenezuelaDataFrame` or even pandas data frames themselves. So, to use those libraries, we would be forces to write unnecessary glue code.

Thus, to tackle the persistence problem, we must decouple the "business" logic (should I call it non-profit logic? _hehe_) from the [cross-cutting concerns](https://en.wikipedia.org/wiki/Cross-cutting_concern) (i.e. persistence); not couple ourselves to a framework--in this case `pandas`.

Below I propose a sample architecture:

```python3
# follows the contract of URL -> DATA (whatever DATA might be, probably yet another data structure)
scraped_data = {
  "url1": { ... }
  "url2": { ... }
}
```

Then, in our wrapper module, we have trivial transformational code:

```python3
# wrappers/pandas.py
import DataFrame from pandas

def scrapedDataAsPandasDataFrame(scraped_data):
    DataFrame.from_dict(scraped_data)
```

Obviously, you cannot put any map into a data frame, so our contract would have to take that into account.

Then, if you NEED data frames for any reason to persist the data, you can just have something like...

```python3
# persistence/whatever.py

class BigQueryRepository:
    def persist(data):
        df = scrapedDataAsPandasDataFrame(data)
        persist(df)
```

Besides separating concerns, the testability of this system is higher, since we would only have functions of signature `Dict -> ?`, instead of `3rdParty -> ?`. If we _must_ couple to something, I'd rather that it is the language itself.

**HAVING SAID ALL OF THAT**

I still agree with @Edilmo in that we should FIRST make it work as fast as you can, and then we can start doing the relevant refactors. That is precisely what I told @LDiazN, to get working code ASAP, and then we can start separating concerns and testing them appropriately. However, it's great that we can discuss all of this **while** Luis is working on the prototype! As the saying goes... _when is the best time to change the code? Before you even wrote it!_ (if you are wondering where that saying came, I think it was from [Raymond H.](https://youtu.be/OSGv2VnC0go?t=931) but I don't recall atm)

>  First, make it possible. Then, make it beautiful. Then, make it fast.
[Source](http://nathanmarz.com/blog/suffering-oriented-programming.html)

Reply:
Made note of this.  It is not complicated to add this conversions. I open an issue #66 to define the scraper interface.

Reply:
I'm going to add the scraper interface design to the tasks board. Will pair on it with @LDiazN possibly tomorrow when we chat. 

@m4r4c00ch0 I agree with your points. A simple map + wrappers will do just fine. I'll work on this with Luis tomorrow 
For now we should aim to just get it working and running in an airflow job, over the test data. That will give us the foundation for everything else + clear up any unknown unknowns associated to hooking it up to airflow.  Once the full thing is working end to end then we can start expanding etc.